### PR TITLE
feat(payment_methods_v2): Payment methods v2 API models

### PR DESCRIPTION
**Pull request by Sarthak1799**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

API models for payment methods v2
- Implemented types for API v2
- Have left a todo! macro for functions where v2 implementation is unclear

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Logic can't be tested as of now as this PR is only concerned with API model change

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/surcharge_decision_configs.rs [hunk 1]
```diff
@@ -403,21 +403,8 @@ pub async fn perform_surcharge_decision_management_for_saved_cards(
...
-                        .parse_enum("CardNetwork")
-                        .change_context(ConfigError::DslExecutionError)
-                })
-                .transpose()?,
-            _ => None,
-        };
-
-        backend_input.payment_method.card_network = card_network;
+        // check in review
+        backend_input.payment_method.card_network = None;
```
Comment (jarnura) [inline crates/router/src/core/payment_methods/surcharge_decision_configs.rs:407]: The code is removed here, is this a expected change?

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/surcharge_decision_configs.rs (hunk 1 above)
Comment (Sarthak1799) [reply crates/router/src/core/payment_methods/surcharge_decision_configs.rs:407]: card_network is removed from the request in v2, so need to decide what to do as it won't be present in here

File: crates/api_models/src/payment_methods.rs [hunk 2]
```diff
@@ -96,6 +103,52 @@ pub struct PaymentMethodCreate {
...
+
+    /// The billing details of the payment method
+    #[schema(value_type = Option<Address>)]
+    pub billing: Option<payments::Address>,
+
+    #[serde(skip_deserializing)]
+    /// The connector mandate details of the payment method, this is added only for cards migration
+    /// api and is skipped during deserialization of the payment method create request as this
+    /// it should not be passed in the request
+    pub connector_mandate_details: Option<PaymentsMandateReference>,
```
[the remark this answers is not in the document]
Comment (Sarthak1799) [reply crates/api_models/src/payment_methods.rs]: Will remove this

File: crates/api_models/src/payment_methods.rs [hunk 3]
```diff
@@ -410,6 +659,52 @@ pub struct PaymentMethodResponse {
...
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[derive(Debug, serde::Deserialize, serde::Serialize, ToSchema, Clone)]
+pub struct PaymentMethodResponse {
+    /// Unique identifier for a merchant
+    #[schema(example = "merchant_1671528864")]
+    pub merchant_id: id_type::MerchantId,
+
+    /// The unique identifier of the customer.
+    #[schema(value_type = Option<String>, max_length = 64, min_length = 1, example = "cus_y3oqhf46pyzuxjbcn2giaqnb44")]
+    pub customer_id: Option<id_type::CustomerId>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:671]: This will a mandatory value

File: crates/api_models/src/payment_methods.rs [hunk 4]
```diff
@@ -901,6 +1323,115 @@ impl<'de> serde::Deserialize<'de> for PaymentMethodListRequest {
...
+    pub card_networks: Option<Vec<api_enums::CardNetwork>>,
+
+    /// Indicates the limit of last used payment methods
+    #[schema(example = 1)]
+    pub limit: Option<i64>,
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+impl<'de> serde::Deserialize<'de> for PaymentMethodListRequest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:1362]: why we need a custom deserialize?

File: crates/router/src/core/payments/helpers.rs [hunk 5]
```diff
@@ -1434,6 +1438,79 @@ pub(crate) async fn get_payment_method_create_request(
     }
 }
 
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[instrument(skip_all)]
+pub(crate) async fn get_payment_method_create_request(
```
Comment (jarnura) [inline crates/router/src/core/payments/helpers.rs:1443]: move this function to payment_method.rs

File: crates/router/src/core/payments/tokenization.rs [hunk 6]
```diff
@@ -729,6 +762,88 @@ async fn skip_saving_card_in_locker(
...
+    api_models::payment_methods::PaymentMethodResponse,
+    Option<payment_methods::transformers::DataDuplicationCheck>,
+)> {
+    let merchant_id = merchant_account.get_id();
+    let customer_id = payment_method_request
+        .clone()
+        .customer_id
+        .clone()
+        .get_required_value("customer_id")?;
+    let payment_method_id = common_utils::generate_id(consts::ID_LENGTH, "pm");
```
Comment (jarnura) [inline crates/router/src/core/payments/tokenization.rs:779]: this function should not create payment_method_id, payment_method needs to be passed here.

File: crates/router/src/core/payments/tokenization.rs [hunk 7]
```diff
@@ -779,6 +894,55 @@ pub async fn save_in_locker(
...
+                &customer_id,
+                merchant_account,
+                None,
+            ))
+            .await
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Add Card Failed")
+        }
+        _ => {
+            let pm_id = common_utils::generate_id(consts::ID_LENGTH, "pm");
```
Comment (jarnura) [inline crates/router/src/core/payments/tokenization.rs:927]: this function should not create payment_method_id, payment_method needs to be passed here.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payment_methods.rs [hunk 8]
```diff
@@ -96,6 +102,72 @@ pub struct PaymentMethodCreate {
     pub network_transaction_id: Option<String>,
 }
 
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone, ToSchema)]
+#[serde(deny_unknown_fields)]
+pub struct PaymentMethodCreate {
+    /// The type of payment method use for the payment.
+    #[schema(value_type = PaymentMethod,example = "card")]
+    pub payment_method: Option<api_enums::PaymentMethod>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:111]:
```suggestion
    pub payment_method: api_enums::PaymentMethod,
```

File: crates/api_models/src/payment_methods.rs [hunk 9]
```diff
@@ -96,6 +102,72 @@ pub struct PaymentMethodCreate {
...
+
+    /// This is a sub-category of payment method.
+    #[schema(value_type = Option<PaymentMethodType>,example = "credit")]
+    pub payment_method_type: Option<api_enums::PaymentMethodType>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:115]:
```suggestion
    pub payment_method_type: api_enums::PaymentMethodType,
```

File: crates/api_models/src/payment_methods.rs [hunk 10]
```diff
@@ -96,6 +102,72 @@ pub struct PaymentMethodCreate {
...
+
+    /// You can specify up to 50 keys, with key names up to 40 characters long and values up to 500 characters long. Metadata is useful for storing additional, structured information on an object.
+    #[schema(value_type = Option<Object>,example = json!({ "city": "NY", "unit": "245" }))]
+    pub metadata: Option<pii::SecretSerdeValue>,
+
+    /// The unique identifier of the customer.
+    #[schema(value_type = Option<String>, max_length = 64, min_length = 1, example = "cus_y3oqhf46pyzuxjbcn2giaqnb44")]
+    pub customer_id: Option<id_type::CustomerId>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:123]:
```suggestion
    pub customer_id: id_type::CustomerId,
```

File: crates/api_models/src/payment_methods.rs [hunk 11]
```diff
@@ -96,6 +102,72 @@ pub struct PaymentMethodCreate {
...
+
+    /// For Client based calls, SDK will use the client_secret
+    /// in order to call /payment_methods
+    /// Client secret will be generated whenever a new
+    /// payment method is created
+    pub client_secret: Option<String>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:129]: Remove this field

File: crates/api_models/src/payment_methods.rs [hunk 12]
```diff
@@ -96,6 +102,72 @@ pub struct PaymentMethodCreate {
...
+
+    /// Payment method data to be passed
+    pub payment_method_data: Option<PaymentMethodCreateData>,
```
Comment (jarnura) [inline crates/api_models/src/payment_methods.rs:132]:
```suggestion
    pub payment_method_data: PaymentMethodCreateData,
```

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by ThisIsMani — DISMISSED** [review]
Auth specific changes looks fine.

**Review by jarnura — APPROVED** [review]

**Review by ThisIsMani — DISMISSED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

File: crates/router/src/core/payment_methods/cards.rs [hunk 13]
```diff
@@ -4726,6 +5023,48 @@ pub async fn get_card_details_with_locker_fallback(
     })
 }
 
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+pub async fn get_card_details_with_locker_fallback(
```
[the remark this answers is not in the document]
Comment (Sarthak1799) [reply crates/router/src/core/payment_methods/cards.rs:5027]: We will anyway take up logic refactoring in a separate PR. We can discuss this as well then.

**Pull request merged**